### PR TITLE
Implement transfer from EVM to Native chain

### PR DESCRIPTION
**Pull request by vldm**

### CODE REVIEW

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+use primitive_types::H160;
+
+use super::*;
+use crate::account_structure::AccountStructure;
+use crate::scope::evm::gweis_to_lamports;
+use solana_sdk::pubkey::Pubkey;
+
+// Currently only static is allowed (but it can be closure).
+type BuiltinImplementation =
+    &'static (dyn Fn(AccountStructure, Vec<Token>, Option<u64>, &Context) -> CallResult + Sync);
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/precompiles/builtins.rs]: O_O

File: evm-utils/evm-bridge/src/main.rs [hunk 2]
```diff
@@ -99,7 +102,36 @@ impl EvmBridge {
             tx.signature.chain_id()
         );
 
-        let ix = solana_evm_loader_program::send_raw_tx(self.key.pubkey(), tx);
+        let mut meta_keys = vec![];
```
Comment (hrls) [inline evm-utils/evm-bridge/src/main.rs:105]: Probably Option instead of Vec

File: evm-utils/evm-state/Cargo.toml [hunk 3]
```diff
@@ -7,7 +7,7 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-evm = { version = "0.23", features = ["with-serde"] }
+evm = { version = "0.23", git = "https://github.com/velas/evm", rev = "53d7b92", features = ["with-serde"] }
```
Comment (hrls) [inline evm-utils/evm-state/Cargo.toml]: What about tag or branch?

File: evm-utils/programs/evm_loader/src/account_structure.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
+use solana_sdk::{keyed_account::KeyedAccount, pubkey::Pubkey};
+#[derive(Copy, Clone, Debug)]
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/account_structure.rs]: Comment lines and then derive

File: evm-utils/programs/evm_loader/src/account_structure.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    /// Returns account of the first user.
+    pub fn user(&self) -> Option<&KeyedAccount> {
+        self.users.get(0)
+    }
+
+    /// Find user by its public key.
+    pub fn find_user(&self, key: &Pubkey) -> Option<&KeyedAccount> {
+        for keyed_account in self.users {
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/account_structure.rs]:
```rust
self.users.iter().find(|keyed| keyed.unsigned_key() == key)
```

File: evm-utils/programs/evm_loader/src/account_structure.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    where
+        F: for<'r> Fn(AccountStructure<'r>) -> U,
+    {
+        use solana_sdk::account::Account;
+        use std::cell::RefCell;
+        let evm_key = Pubkey::new_unique();
+        let evm_account = RefCell::new(crate::create_state_account());
+        let evm_state = KeyedAccount::new(&evm_key, false, &evm_account);
+
+        let keys: Vec<_> = (0..(num_keys + 1))
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/account_structure.rs]: `std::iter::repeat_with(|| ...).take(num_keys).collect()`

File: evm-utils/programs/evm_loader/src/account_structure.rs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    /// Create AccountStructure for testing purposes, with random accounts.
+    #[cfg(test)]
+    pub(crate) fn testing<F, U>(num_keys: usize, func: F) -> U
+    where
+        F: for<'r> Fn(AccountStructure<'r>) -> U,
+    {
+        use solana_sdk::account::Account;
+        use std::cell::RefCell;
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/account_structure.rs:39]: `\n`

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        implementation: BuiltinImplementation,
+    ) -> Self {
+        assert_eq!(function_hash.len(), 4);
+        Self {
+            function_hash,
+            abi,
+            implementation,
+        }
+    }
+    pub fn parse_abi(
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/precompiles/builtins.rs]: Why this isn't part of Function? If `abi: Function` then why `Result::Ok : Vec<Token>`?

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+pub fn eth_to_sol_parse_inputs(inputs: Vec<Token>) -> Result<Pubkey, super::PrecompileErrors> {
+    ensure!(
+        inputs.len() == 1,
+        ParamsCountMismatch {
+            expected: 1_usize,
+            got: inputs.len()
+        }
+    );
+
+    let bytes = match &inputs[0] {
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/precompiles/builtins.rs]: `slice::first`

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+use solana_sdk::pubkey::Pubkey;
+
+// Currently only static is allowed (but it can be closure).
+type BuiltinImplementation =
+    &'static (dyn Fn(AccountStructure, Vec<Token>, Option<u64>, &Context) -> CallResult + Sync);
+
+#[derive(Clone)]
+pub struct Builtin {
+    // TODO: Replace by real function hash calculation
+    function_hash: Vec<u8>,
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/precompiles/builtins.rs]: `[u8; 4]`

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs (hunk 9 above)
Comment (vldm) [reply evm-utils/programs/evm_loader/src/precompiles/builtins.rs]: 0, 1 is allowed constants.

`.first().unwrap()` is more verbose to choose it for some reasons.

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs (hunk 8 above)
Comment (vldm) [reply evm-utils/programs/evm_loader/src/precompiles/builtins.rs]: ethabi::Function is extern interface.

File: evm-utils/programs/evm_loader/src/precompiles/builtins.rs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+use primitive_types::H160;
+
+use super::abi_parse::ParseTokens;
+use super::errors::*;
+use super::{CallResult, PrecompileContext, PrecompileOk, Result};
+use crate::scope::evm::gweis_to_lamports;
+use solana_sdk::pubkey::Pubkey;
+
+pub trait BuiltinFunction<Inputs> {
+    fn call(&self, inputs: Inputs, cx: PrecompileContext<'_>) -> Result<PrecompileOk>;
```
Comment (hrls) [inline evm-utils/programs/evm_loader/src/precompiles/builtins.rs:16]: Usually Context is first arg.
`func.call(Args::parse_from(|bytes| { ... }).map(SomeArg), ctx))` vs `func.call(ctx, Args::parse_from(|bytes| { ... }).map(SomeArg))`. Especially when args parsing is multiline code.

**Review by hrls — APPROVED** [review]

**Pull request merged**